### PR TITLE
Feature/refactor streaming

### LINKED ISSUE #364: Refactor the stream API
### Is your feature request related to a problem? Please describe.

The proposal is to change the usage of streams in the following way.

When `stream=True` the returned object would be a Fieldlist (for GRIB data):
```python
ds = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds:
     # f is now a Field

# at this point ds consumed the stream
```

Iterating in batches would be a generic option (not only stream specific):

```python
ds1 = from_source("file", "my_local_data.grib")
ds2 = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds1.batched(2):
     # f is now a Fieldlist with 2 Fields

for f in ds2.batched(2):
     # f is now a Fieldlist with 2 Fields
```
`group_by` would behave in a similar way. 

```python
ds1 = from_source("file", "my_local_data.grib")
ds2 = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds1.group_by("level"):
     # f is now a Fieldlist

for f in ds2.group_by("level"):
     # f is now a Fieldlist
```

Please note that using `group_by` for non-stream data will be based on the metadata from the full dataset. However, for the stream it would be simply built by consuming GRIB messages from the stream until the values of the metadata keys specified in `group_by` change. 

We could read the whole stream into memory with the `read_all` option:

```python
ds = from_source("url", "http://..../my_data.grib", stream=True, read_all=True)

# ds is now a Fieldlist in memory, so all these work
len(ds)
r = ds.sel(param="t")

for f in ds:
     # f is now a Field

for f in ds.batched(2):
     # f is now a Fieldlist with 2 Fields
```

### PR DESCRIPTION
**Pull request by sandorkertesz**
Fixes #364 

This PR:
-  adds`batched()` and `group_by()` for stream and fieldlist like objects
-  adds the`read_all` option for `from_source()` when `stream=True`
-  removes the `batch_size` and `group_by` options from `from_source` (This is a breaking change!)

Examples
-----------

```python
ds = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds:
     # f is a field

# at this point ds consumed the stream
```

```python
ds1 = from_source("file", "my_local_data.grib")
ds2 = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds1.batched(2):
     # f is now a Fieldlist with 2 Fields

for f in ds2.batched(2):
     # f is now a Fieldlist with 2 Fields
```

```python
ds1 = from_source("file", "my_local_data.grib")
ds2 = from_source("url", "http://..../my_data.grib", stream=True)

for f in ds1.group_by("level"):
     # f is a Fieldlist

for f in ds2.group_by("level"):
     # f is a Fieldlist
```

```python
ds = from_source("url", "http://..../my_data.grib", stream=True, read_all=True)

# ds is a Fieldlist in memory, so all these work
len(ds)
r = ds.sel(param="t")

for f in ds:
     # f isa Field

for f in ds.batched(2):
     # f is a Fieldlist with 2 Fields
```

Documentation
-----------------
- `batched`: https://earthkit-data--371.org.readthedocs.build/en/371/_api/data/core/fieldlist/index.html#data.core.fieldlist.FieldList.batched
- `group_by` https://earthkit-data--371.org.readthedocs.build/en/371/_api/data/core/fieldlist/index.html#data.core.fieldlist.FieldList.group_by
- streams: https://earthkit-data--371.org.readthedocs.build/en/371/guide/streams.html

Notebooks
------------
https://earthkit-data--371.org.readthedocs.build/en/371/examples/data_from_stream.html
https://earthkit-data--371.org.readthedocs.build/en/371/examples/url_stream.html
https://earthkit-data--371.org.readthedocs.build/en/371/examples/fdb.html

### CODE REVIEW

File: earthkit/data/core/__init__.py [hunk 1]
```diff
@@ -152,3 +152,11 @@ def _not_implemented(self):
         if hasattr(self, "path"):
             extra = f" on {self.path}"
         raise NotImplementedError(f"{module}.{name}.{func}(){extra}")
+
+    def batched(self, *args):
```
Comment (iainrussell) [inline earthkit/data/core/__init__.py:156]: I wonder if this should be batch() so that it is a verb, like group_by()?

**Review by iainrussell — CHANGES_REQUESTED** [review]
Apart from that suggestion (what do you think @sandorkertesz?), I think this looks really really nice - a good and more intuitive way of doing things.

File: earthkit/data/core/__init__.py (hunk 1 above)
Comment (iainrussell) [reply earthkit/data/core/__init__.py:156]: That's a good reason!

**Review by tlmquintino — APPROVED** [review]
Looks good. Thanks for clearing up the interface.

**Pull request merged**